Try to parse all valid Testkit HOCON comments

mdedetrich opened the pull request:
Resolves: https://github.com/scalacenter/scalafix/issues/1828

Pinging @bjaglin since you commented on the original issue. The only thing that is missing is a test, its my first time working on scalafix so not entirely sure where to put such a test and what type of test.

bjaglin commented on scalafix-testkit/src/main/scala/scalafix/testkit/RuleTest.scala:
```diff
@@ -31,36 +30,48 @@ object RuleTest {
+          val rules = decoder.read(rulesConf).get.withConfiguration(config).get
+          (rules, sdoc)
+      } match {
+        case None =>
```
it looks like you forgot to unpack the happy path

mdedetrich replied:
The idea is that we ignore the unhappy path, i.e. we only care about the first success case which occurs on `collectFirst`.

Or am I missing something?

bjaglin commented on scalafix-testkit/src/main/scala/scalafix/testkit/RuleTest.scala:
```diff
@@ -31,36 +30,48 @@ object RuleTest {
+
+      lazilyParsedComments.collectFirst {
+        case Some((decoder, conf, sdoc)) =>
+          val rulesConf = ConfGet
+            .getKey(conf, "rules" :: "rule" :: Nil)
+            .getOrElse(Conf.Lst(Nil))
```
per the message in the exception, this should be lifted up to the `lazilyParsedComments` closure to effectively skip comments that are HOCON but do not contain `rule` or `rules`

mdedetrich replied:
So I thought the failure in parsing is done in `Conf.parseString(test.testName, syntax)` (which is why I pattern match against it). The idea behind my solution is that we create a lazy collection called `lazilyParsedComments` of successfully/unsuccessfully parsed comments and then we do `collectFirst` to get the first successful parsing.

In other words, the logic of figuring out what is a correct HOCON comment has already happened by the time we get to `rulesConf`

bjaglin replied:
> This is definitely an improvement and should cover your original use-case of a long comment

Actually, it does not - the test I added in fd35cf2b6e63fb251e73a152a0badede1b93ab75 shows it. I will try to do what I was suggesting in another commit and see if it helps.

bjaglin commented on scalafix-testkit/src/main/scala/scalafix/testkit/RuleTest.scala:
```diff
@@ -31,36 +30,48 @@ object RuleTest {
+          (rules, sdoc)
+      } match {
+        case None =>
+          throw new IllegalArgumentException("Expected a single comment with a valid rule or rules key defined")
```
adding the context (path in `test`) would help here

mdedetrich replied:
Done

bjaglin commented on scalafix-testkit/src/main/scala/scalafix/testkit/RuleTest.scala:
```diff
@@ -2,8 +2,7 @@ package scalafix.testkit
 import scala.meta.internal.symtab.SymbolTable
-
-import metaconfig.Conf
+import metaconfig.{Conf, Configured}
```
make sure to run the [formatters](https://github.com/scalacenter/scalafix/blob/main/CONTRIBUTING.md#formatting) to please the CI

mdedetrich replied:
Done

bjaglin reviewed (commented):
Thanks for the PR!

For testing, I would suggest an integration test, simply copy/pasting https://github.com/scalacenter/scalafix/blob/main/scalafix-tests/input/src/main/scala-2/test/LintAsserts.scala (the `Disable` rule is useless though) in its current directory, and adding a license header.

mdedetrich commented:
@bjaglin Thanks for reviewing the PR, I addressed some of the comments but I may have an incorrect understanding of what the problem is since it appears that `RuleSuite` is not passing, investigating it more

mdedetrich commented:
@bjaglin Okay so I have done some improvements to the PR, on my local machine `sbt testOnly scalafix.tests.rule.RuleSuite` is passing, some other tests are failing but that appears to be unrelated (jgit/git/signing related stuff).

Would it be possible to re-run the CI just to confirm I didn't create any regressions?

bjaglin commented:
> Would it be possible to re-run the CI just to confirm I didn't create any regressions?

Done. I also changed [the default](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#controlling-changes-from-forks-to-workflows-in-public-repositories), so I shouldn't need to approve execution at next push :crossed_fingers:

bjaglin reviewed (commented):
You can commit the test file and run the formatters and we should be good to go :+1:

mdedetrich commented:
@bjaglin Thanks for helping out, I will try and have a look at why the test is failing in the next few days

mdedetrich commented on scalafix-testkit/src/main/scala/scalafix/testkit/RuleTest.scala:
```diff
@@ -31,36 +33,54 @@ object RuleTest {
+        val syntax = comment.syntax.stripPrefix("/*").stripSuffix("*/")
+        for {
+          conf <- Try(Conf.parseString(test.testName, syntax)).toEither
+            .flatMap(_.toEither)
```
@bjaglin I didn't realize you were working on this right now, thanks for fixing the bug!

I think there is a slight improvement that can be done here, this expression ends up getting the type `Either[Serializable, _]` because the first `toEither` has `Throwable` on the left where as the second has `ConfError`, should I fix this?

bjaglin replied:
I am planning to cut a release tonight ahead of the official release of scala 2.13.12, that's why I wanted to get this out ASAP.

As for the type, I am not sure it really matters since we never unpack the left? I'll merge this if it's green in an hour or so, feel free to push if you see further improvements.

mdedetrich replied:
I already have the fix written, its quite simple. Just running scalafmt/scalafix locally and ill push immediately after

mdedetrich commented:
@bjaglin Just pushed the fix I was talking about, tests are passing locally so all should be good.

bjaglin commented:
Thanks @mdedetrich !
